sdk/core/switcher/entry.S

```diff
@@ -79,23 +79,47 @@ switcher_scheduler_entry_csp:
 	forall reloadOne, \reg1, \regs
 .endm
 
+/**
+ * Verify the compartment stack is valid, with the expected permissions and
+ * unsealed.
+ * This macro assumes t2 and tp are available to use.
+ */
+.macro check_compartment_stack_integrity reg
+	// make sure the caller's CSP is valid
+	cgettag         t2, \reg
+	beqz            t2, .Lforce_unwind
+	// make sure the caller's CSP has the expected permissions
+	cgetperm        t2, \reg
+	li              tp, COMPARTMENT_STACK_PERMISSIONS
+	bne	            tp, t2, .Lforce_unwind
+	// make sure the caller's CSP is unsealed
+	cgettype        t2, \reg
+	bnez            t2, .Lforce_unwind
+.endm
+
 	.section .text, "ax", @progbits
 	.globl compartment_switcher_entry
 	.p2align 2
 	.type compartment_switcher_entry,@function
 compartment_switcher_entry:
+	// before we access any privileged state, we can verify the
+	// compartment's csp is valid. If no, force unwind.
+	check_compartment_stack_integrity csp
 	// The caller should back up all callee saved registers.
 	// mscratchc should always have an offset of 0.
 	cspecialr		ct2, mscratchc
 #ifndef NDEBUG
 	// XXX: This line is useless, only for mscratch to show up in debugging.
 	cmove			ct2, ct2
 #endif
-	clhu			tp, TrustedStack_offset_frameoffset(ct2)
+
 	// make sure the trusted stack is still in bounds
-	cgetlen                 t2, ct2
-	bgeu                    tp, t2, .Lout_of_trusted_stack
+	clhu			tp, TrustedStack_offset_frameoffset(ct2)
```

Comment:
Can you fix indenting? It looks like you're mixing tabs and spaces.

Reply:
good catch, thanks :)